sql/core/src/main/scala/org/apache/spark/sql/execution/RemoveRedundantProjects.scala

```diff
@@ -58,7 +58,8 @@ object RemoveRedundantProjects extends Rule[SparkPlan] {
           p.mapChildren(removeProject(_, false))
         }
       case op: TakeOrderedAndProjectExec =>
-        op.mapChildren(removeProject(_, false))
+        // TakeOrderedAndProjectExec requires keep column ordering if AQE is enabled.
```

Comment:
Let's be more specific about it. AQE is not the direct reason. The root cause is the planner turns `Limit` + `Sort` into `TakeOrderedAndProjectExec` which adds an additional projection that does not exist in the logical plan. We shouldn't use this additional projection to optimize out other projections, otherwise when AQE turns physical plan back to logical plan, we lose the `Project` and may mess up the output column order.

So the condition shouldn't be `conf.adaptiveExecutionEnabled`, but `op.projectList == op.child.output`

Reply:
When AQE is disabled, using `op.projectList == op.child.output` will affect the plan.
Before:
```
TakeOrderedAndProject(limit=100, orderBy=[year#271 ASC NULLS FIRST,course#270 ASC NULLS FIRST], output=[year#271,course#270,earnings#272,balance#269])
+- Window [sum(earnings#272) windowspecdefinition(year#271 ASC NULLS FIRST, course#270 ASC NULLS FIRST, specifiedwindowframe(RangeFrame, unboundedpreceding$(), currentrow$())) AS balance#269], [year#271 ASC NULLS FIRST, course#270 ASC NULLS FIRST]
   +- *(2) Sort [year#271 ASC NULLS FIRST, course#270 ASC NULLS FIRST], false, 0
      +- Exchange SinglePartition, ENSURE_REQUIREMENTS, [plan_id=122]
         +- *(1) ColumnarToRow
            +- FileScan parquet spark_catalog.default.t[course#270,year#271,earnings#272]
```
After:
```
TakeOrderedAndProject(limit=100, orderBy=[year#271 ASC NULLS FIRST,course#270 ASC NULLS FIRST], output=[year#271,course#270,earnings#272,balance#269])
+- Window [sum(earnings#272) windowspecdefinition(year#271 ASC NULLS FIRST, course#270 ASC NULLS FIRST, specifiedwindowframe(RangeFrame, unboundedpreceding$(), currentrow$())) AS balance#269], [year#271 ASC NULLS FIRST, course#270 ASC NULLS FIRST]
   +- *(2) Sort [year#271 ASC NULLS FIRST, course#270 ASC NULLS FIRST], false, 0
      +- Exchange SinglePartition, ENSURE_REQUIREMENTS, [plan_id=122]
         +- *(1) Project [year#271, course#270, earnings#272]
            +- *(1) ColumnarToRow
               +- FileScan parquet spark_catalog.default.t[course#270,year#271,earnings#272]
```

Reply:
ok, how about `aqeEnabled && op.projectList == op.child.output`?

Reply:
That's great.